modules/xfem/src/userobjects/InterfaceMeshCut3DUserObject.C

```diff
@@ -0,0 +1,243 @@
+
+      for (const auto & node_neigh_elem_id : _node_to_elem_map[node.id()])
+      {
+        const Elem & node_neigh_elem = _cutter_mesh->elem_ref(node_neigh_elem_id);
+        std::vector<Point> vertices{
+            node_neigh_elem.node_ref(0), node_neigh_elem.node_ref(1), node_neigh_elem.node_ref(2)};
+        Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+        unsigned int j = node_neigh_elem.local_node(node.id());
+        Point normal_at_node_j = elem_plane.unit_normal(vertices[0]);
+        unsigned int m = j + 1 < 3 ? j + 1 : j + 1 - 3;
+        unsigned int n = j + 2 < 3 ? j + 2 : j + 2 - 3;
+        Point line_1 = node_neigh_elem.node_ref(j) - node_neigh_elem.node_ref(m);
+        Point line_2 = node_neigh_elem.node_ref(j) - node_neigh_elem.node_ref(n);
+        Real dot = line_1 * line_2;
+        Real lenSq1 = line_1 * line_1;
+        Real lenSq2 = line_2 * line_2;
+        Real angle = std::acos(dot / std::sqrt(lenSq1 * lenSq2));
+        normal_at_node += normal_at_node_j * angle;
+        angle_sum += angle;
+      }
+      normal[1 + i] = normal_at_node;
+    }
+
+    for (unsigned int i = 0; i < elem->n_sides(); i++)
+    {
+      std::vector<Point> vertices{elem->node_ref(0), elem->node_ref(1), elem->node_ref(2)};
+
+      Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+      Point normal_at_edge = libMesh::pi * elem_plane.unit_normal(vertices[0]);
+
+      const Elem * neighbor = elem->neighbor_ptr(i);
+
+      if (neighbor != nullptr)
+      {
+        std::vector<Point> vertices{
+            neighbor->node_ref(0), neighbor->node_ref(1), neighbor->node_ref(2)};
+
+        Plane elem_plane(vertices[0], vertices[1], vertices[2]);
+        normal_at_edge += libMesh::pi * elem_plane.unit_normal(vertices[0]);
+      }
+      normal[4 + i] = normal_at_edge;
+    }
+    _pseudo_normal.insert(std::make_pair(elem->id(), normal));
+  }
+}
+
+Point
+InterfaceMeshCut3DUserObject::nodeNormal(const unsigned int & node_id)
+{
+  Point normal(0.0);
+
+  for (const auto & node_neigh_elem_id : _node_to_elem_map[node_id])
+  {
+    const auto & elem = _cutter_mesh->elem_ref(node_neigh_elem_id);
+    Plane elem_plane(elem.node_ref(0), elem.node_ref(1), elem.node_ref(2));
+    normal += elem_plane.unit_normal(elem.node_ref(0));
+  }
+
+  unsigned int num = _node_to_elem_map[node_id].size();
+  return normal / num;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * /*elem*/,
+                                                   std::vector<Xfem::CutEdge> & /*cut_edges*/,
+                                                   std::vector<Xfem::CutNode> & /*cut_nodes*/) const
+{
+  mooseError("invalid method for 3D mesh cutting");
+  return false;
+}
+
+bool
+InterfaceMeshCut3DUserObject::cutElementByGeometry(const Elem * elem,
+                                                   std::vector<Xfem::CutFace> & cut_faces) const
+{
+  bool elem_cut = false;
+
+  if (elem->dim() != 3)
```

Comment:
This should be an assert since if this failed, this would be an error from the code that called it. We should probably have this at the beginning of all of our `cutElementByGeometry()` methods (both for the 2D and 3D ones).

I would suggest doing this: Put this generically worded assert (or the 2D equivalent of it) at the beginning of every `cutElementByGeometry()` method we have:
```
mooseAssert(elem-Dim() == 3, "Dimension of element to be cut must be 3");